pkg/cluster/bootstrap.go

```diff
@@ -85,47 +90,219 @@ func (c *Cluster) shouldBootstrapLoad(ctx context.Context) (bool, error) {
+			Content: v,
+		}
+	}
+
+	return sc.Update(ctx, storageKey(normalizedToken), rev, data)
+}
+
+// ReconcileBootstrapData is called before any data is saved to the
+// datastore or locally. It checks to see if the contents of the
+// bootstrap data in the datastore is newer than on disk or different
+//  and dependingon where the difference is, the newer data is written
+// to the older.
+func (c *Cluster) ReconcileBootstrapData(ctx context.Context, crb *config.ControlRuntimeBootstrap) error {
+	logrus.Info("Reconciling bootstrap data between database and disk")
+
+	storageClient, err := client.New(c.etcdConfig)
+	if err != nil {
+		return err
+	}
+
+	token := c.config.Token
+	if token == "" {
+		tokenFromFile, err := readTokenFromFile(c.runtime.ServerToken, c.runtime.ServerCA, c.config.DataDir)
+		if err != nil {
+			return err
+		}
+		if tokenFromFile == "" {
+			// at this point this is a fresh start in a non-managed environment
+			c.saveBootstrap = true
+			return nil
+		}
+		token = tokenFromFile
+	}
+	normalizedToken, err := normalizeToken(token)
+	if err != nil {
+		return err
+	}
+
+	var value *client.Value
+
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for range ticker.C {
+		value, err = c.getBootstrapKeyFromStorage(ctx, storageClient, normalizedToken, token)
+		if err != nil {
+			if strings.Contains(err.Error(), "not supported for learner") { // create error value
+				continue
+			}
+			return err
+		}
+		if value == nil {
+			return nil
+		}
+
+		break
+	}
+
+	data, err := decrypt(normalizedToken, value.Data)
+	if err != nil {
 		return err
 	}
 
-	// return if file already exists
-	if _, err := os.Stat(stamp); err == nil {
-		return nil
+	if err := c.certDirsExist(); err != nil {
+		logrus.Warn(err.Error())
+		return bootstrap.WriteToDiskFromStorage(bytes.NewBuffer(data), crb)
 	}
 
-	// otherwise try to create it
-	f, err := os.Create(stamp)
+	paths, err := bootstrap.ObjToMap(crb)
 	if err != nil {
 		return err
 	}
 
-	return f.Close()
+	files := make(bootstrap.DataFormat)
+	if err := json.NewDecoder(bytes.NewBuffer(data)).Decode(&files); err != nil {
+		// This will fail if data is being pulled from old an cluster since
+		// older clusters used a map[string][]byte for the data structure.
+		// Therefore, we need to perform a migration to the newer bootstrap
+		// format; bootstrap.BootstrapFile.
+		if err := c.migrateBootstrapData(ctx, storageClient, data, files, normalizedToken, value.Modified); err != nil {
+			return err
+		}
+	}
+
+	for pathKey, fileData := range files {
+		path, ok := paths[pathKey]
+		if !ok {
+			continue
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		fd, err := ioutil.ReadAll(f)
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(fileData.Content, fd) {
+			logrus.Warnf("%s is out of sync with datastore", path)
+
+			info, err := os.Stat(path)
+			if err != nil {
+				return err
+			}
+
+			switch {
```

Comment:
I feel like we need to allow for a couple seconds of clock skew in here, otherwise it's going to be very twitchy.

Reply:
What scenarios do you see playing out otherwise?

Reply:
NTP being off by a few seconds since a node just started or restarted, golang Time precision being higher than filesystem precision, etc.

Reply:
That seems like that'd be out of scope for this to worry about and be more of an operator concern at system level.

Reply:
Yeah, but is there any reason not to guard against it? Just change the check to `info.ModTime().Unix() - files[pathKey].Timestamp.Unix() >= 5 * time.Second` or whatever skew constant we want to use.

Reply:
updated